Redid handling of link down to ensure tear down of old path

Ktmi opened the pull request:
Closes #517, #575

### Summary

Replaces the procedure for handling of link down events to ensure that the old_path is torn down before fully committing changes to the database.

### Local Tests

This still needs some tweaking, but the general shape of the solution is here. Will have to develop a test harness which can replicate the old issue reliably.

### E2E Tests

Here's the latest E2E test report:

```
kytos-1  | Starting enhanced syslogd: rsyslogd.
kytos-1  | /etc/openvswitch/conf.db does not exist ... (warning).
kytos-1  | Creating empty database /etc/openvswitch/conf.db.
kytos-1  | Starting ovsdb-server.
kytos-1  | rsyslogd: error during config processing: omfile: chown for file '/var/log/syslog' failed: Operation not permitted [v8.2302.0 try https://www.rsyslog.com/e/2207 ]
kytos-1  | Configuring Open vSwitch system IDs.
kytos-1  | Starting ovs-vswitchd.
kytos-1  | Enabling remote OVSDB managers.
kytos-1  | There is no NAPPS_PATH specified. Default will be used.
kytos-1  | + '[' -z '' ']'
kytos-1  | + '[' -z '' ']'
kytos-1  | + echo 'There is no NAPPS_PATH specified. Default will be used.'
kytos-1  | + NAPPS_PATH=
kytos-1  | + sed -i 's/STATS_INTERVAL = 60/STATS_INTERVAL = 7/g' /var/lib/kytos/napps/kytos/of_core/settings.py
kytos-1  | + sed -i 's/CONSISTENCY_MIN_VERDICT_INTERVAL =.*/CONSISTENCY_MIN_VERDICT_INTERVAL = 60/g' /var/lib/kytos/napps/kytos/flow_manager/settings.py
kytos-1  | + sed -i 's/LINK_UP_TIMER = 10/LINK_UP_TIMER = 1/g' /var/lib/kytos/napps/kytos/topology/settings.py
kytos-1  | + sed -i 's/DEPLOY_EVCS_INTERVAL = 60/DEPLOY_EVCS_INTERVAL = 5/g' /var/lib/kytos/napps/kytos/mef_eline/settings.py
kytos-1  | + sed -i 's/LLDP_LOOP_ACTIONS = \["log"\]/LLDP_LOOP_ACTIONS = \["disable","log"\]/' /var/lib/kytos/napps/kytos/of_lldp/settings.py
kytos-1  | + sed -i 's/LLDP_IGNORED_LOOPS = {}/LLDP_IGNORED_LOOPS = {"00:00:00:00:00:00:00:01": \[\[4, 5\]\]}/' /var/lib/kytos/napps/kytos/of_lldp/settings.py
kytos-1  | + sed -i 's/CONSISTENCY_COOKIE_IGNORED_RANGE =.*/CONSISTENCY_COOKIE_IGNORED_RANGE = [(0xdd00000000000000, 0xdd00000000000009)]/g' /var/lib/kytos/napps/kytos/flow_manager/settings.py
kytos-1  | + sed -i 's/LIVENESS_DEAD_MULTIPLIER =.*/LIVENESS_DEAD_MULTIPLIER = 3/g' /var/lib/kytos/napps/kytos/of_lldp/settings.py
kytos-1  | + kytosd --help
kytos-1  | + sed -i s/WARNING/INFO/g /etc/kytos/logging.ini
kytos-1  | + test -z ''
kytos-1  | + TESTS=tests/
kytos-1  | + test -z ''
kytos-1  | + RERUNS=2
kytos-1  | + python3 scripts/wait_for_mongo.py
kytos-1  | Trying to run hello command on MongoDB...
kytos-1  | Trying to run 'hello' command on MongoDB...
kytos-1  | Trying to run 'hello' command on MongoDB...
kytos-1  | Ran 'hello' command on MongoDB successfully. It's ready!
kytos-1  | + python3 -m pytest tests/ --reruns 2 -r fEr
kytos-1  | ============================= test session starts ==============================
kytos-1  | platform linux -- Python 3.11.2, pytest-8.1.1, pluggy-1.5.0
kytos-1  | rootdir: /tests
kytos-1  | plugins: rerunfailures-13.0, timeout-2.2.0, anyio-4.3.0
kytos-1  | collected 279 items
kytos-1  | 
kytos-1  | tests/test_e2e_01_kytos_startup.py ..                                    [  0%]
kytos-1  | tests/test_e2e_05_topology.py ..................                         [  7%]
kytos-1  | tests/test_e2e_06_topology.py ....                                       [  8%]
kytos-1  | tests/test_e2e_10_mef_eline.py ..........ss.....x.........RRF........... [ 22%]
kytos-1  | .RRFRRF                                                                  [ 23%]
kytos-1  | tests/test_e2e_11_mef_eline.py ......                                    [ 25%]
kytos-1  | tests/test_e2e_12_mef_eline.py .....Xx.                                  [ 28%]
kytos-1  | tests/test_e2e_13_mef_eline.py ....Xs.s.....Xs.s.XXxX.xxxx..X........... [ 43%]
kytos-1  | .                                                                        [ 43%]
kytos-1  | tests/test_e2e_14_mef_eline.py xRRFRRFRRF                                [ 45%]
kytos-1  | tests/test_e2e_15_mef_eline.py .....                                     [ 46%]
kytos-1  | tests/test_e2e_16_mef_eline.py ..                                        [ 47%]
kytos-1  | tests/test_e2e_17_mef_eline.py ...                                       [ 48%]
kytos-1  | tests/test_e2e_20_flow_manager.py ..........................             [ 58%]
kytos-1  | tests/test_e2e_21_flow_manager.py ...                                    [ 59%]
kytos-1  | tests/test_e2e_22_flow_manager.py ...............                        [ 64%]
kytos-1  | tests/test_e2e_23_flow_manager.py ..............                         [ 69%]
kytos-1  | tests/test_e2e_30_of_lldp.py .R...                                       [ 70%]
kytos-1  | tests/test_e2e_31_of_lldp.py ...                                         [ 72%]
kytos-1  | tests/test_e2e_32_of_lldp.py ...                                         [ 73%]
kytos-1  | tests/test_e2e_40_sdntrace.py ................                           [ 78%]
kytos-1  | tests/test_e2e_41_kytos_auth.py ........                                 [ 81%]
kytos-1  | tests/test_e2e_42_sdntrace.py ..                                         [ 82%]
kytos-1  | tests/test_e2e_50_maintenance.py ............................            [ 92%]
kytos-1  | tests/test_e2e_60_of_multi_table.py .....                                [ 94%]
kytos-1  | tests/test_e2e_70_kytos_stats.py ........                                [ 97%]
kytos-1  | tests/test_e2e_80_pathfinder.py ss......                                 [100%]
kytos-1  | 
kytos-1  | =================================== FAILURES ===================================
...
kytos-1  | =========================== rerun test summary info ============================
kytos-1  | RERUN test_e2e_10_mef_eline.py::TestE2EMefEline::test_150_current_path_value_given_dynamic_backup_path_and_empty_primary_conditions
kytos-1  | RERUN test_e2e_10_mef_eline.py::TestE2EMefEline::test_150_current_path_value_given_dynamic_backup_path_and_empty_primary_conditions
kytos-1  | RERUN test_e2e_10_mef_eline.py::TestE2EMefEline::test_300_inter_evc_dynamic_uni_status
kytos-1  | RERUN test_e2e_10_mef_eline.py::TestE2EMefEline::test_300_inter_evc_dynamic_uni_status
kytos-1  | RERUN test_e2e_10_mef_eline.py::TestE2EMefEline::test_301_inter_evc_static_uni_status
kytos-1  | RERUN test_e2e_10_mef_eline.py::TestE2EMefEline::test_301_inter_evc_static_uni_status
kytos-1  | RERUN test_e2e_14_mef_eline.py::TestE2EMefEline::test_010_redeploy_avoid_vlan
kytos-1  | RERUN test_e2e_14_mef_eline.py::TestE2EMefEline::test_010_redeploy_avoid_vlan
kytos-1  | RERUN test_e2e_14_mef_eline.py::TestE2EMefEline::test_015_redeploy_avoid_primary_path
kytos-1  | RERUN test_e2e_14_mef_eline.py::TestE2EMefEline::test_015_redeploy_avoid_primary_path
kytos-1  | RERUN test_e2e_14_mef_eline.py::TestE2EMefEline::test_020_redeploy_avoid_vlan_static_path
kytos-1  | RERUN test_e2e_14_mef_eline.py::TestE2EMefEline::test_020_redeploy_avoid_vlan_static_path
kytos-1  | RERUN test_e2e_30_of_lldp.py::TestE2EOfLLDP::test_010_disable_of_lldp
kytos-1  | =========================== short test summary info ============================
kytos-1  | FAILED tests/test_e2e_10_mef_eline.py::TestE2EMefEline::test_150_current_path_value_given_dynamic_backup_path_and_empty_primary_conditions
kytos-1  | FAILED tests/test_e2e_10_mef_eline.py::TestE2EMefEline::test_300_inter_evc_dynamic_uni_status
kytos-1  | FAILED tests/test_e2e_10_mef_eline.py::TestE2EMefEline::test_301_inter_evc_static_uni_status
kytos-1  | FAILED tests/test_e2e_14_mef_eline.py::TestE2EMefEline::test_010_redeploy_avoid_vlan
kytos-1  | FAILED tests/test_e2e_14_mef_eline.py::TestE2EMefEline::test_015_redeploy_avoid_primary_path
kytos-1  | FAILED tests/test_e2e_14_mef_eline.py::TestE2EMefEline::test_020_redeploy_avoid_vlan_static_path
kytos-1  | = 6 failed, 250 passed, 8 skipped, 8 xfailed, 7 xpassed, 1369 warnings, 13 rerun in 11466.55s (3:11:06) =

[Kkytos-1 exited with code 1

```

viniarck commented on main.py:
```diff
@@ -829,79 +830,159 @@ def handle_link_down(self, event):
-                    # tradditional way
-                    if (
+        with ExitStack() as exit_stack:
+            exit_stack.enter_context(self._lock)
```
`handle_link_down` will always be executed single threaded, `self._lock` was meant only for consistency check related execute loop, I wouldn't recommend mixing it here, so until we get to consistency part I wouldn't even touch this. 

Each `evc.lock` though yes, those ones we need to make sure that when acquired by this single threaded exec no other concurrent threads would be performing other side effects. The `ExitStack()` usage is great, if it'll be used, needs to also ensure that deletions are also executed atomically. So, if you can review this.

viniarck replied:
In summary: all affected EVCs need to build/map the flows and perform the flows removals and installations with the bulk request that we discussed in the issue.

Ktmi replied:
Alright, I'll change it to a separate lock. A lock is necessary here if we where to have multiple threads that wanted to hold multiple EVC locks, so I'll add one that is named explicitly for such purposes.

viniarck commented on main.py:
```diff
@@ -829,79 +830,159 @@ def handle_link_down(self, event):
-                    ):
-                        evcs_normal.append(evc)
+                    )):
+                        redeploy.append(evc)
```
if `evc.is_affected_by_link(link)` is True 
and `evc.is_failover_path_affected_by_link(link)` is True

failover_path needs to be cleaned too. Otherwise this can end up keeping a failover_path that isn't UP. Can you review this conditional and test it?

Ktmi replied:
Well, that's what the old logic was saying to do here. But I do see your point. Looking at the original code, it wouldn't ever tear down the failover path. Just tear down, then setup the current path and try to deploy a failover path if it doesn't already exist.

viniarck replied:
Yes, that was another adjacent bug in the logic. It's the [item 1.2](https://github.com/kytos-ng/mef_eline/issues/517#issuecomment-2511372895). It would leave `failover_path` not UP (with a link down in the path), and then a subsequent link_down in the `current_path` would switchover to this `failover_path` that isn't UP

viniarck reviewed (commented):
@Ktmi I read in the description your still working on it and with TODOs in the code. Overall, it's headed in the right direction. 

Yes, first and foremost we definitely need a local way of reproducing this either deterministically or with a high likelihood to assess with concrete data how it's performing in terms of correctness and failover convergence control plane and data plane switchover flows performance, and e2e test maybe we don't need fully exact one but one that simulates a closely double failure, and other cases like failing both current_paht and failover_path when they're not fully disjoint. 

Other than that, when you need a final review let me know, I won't completely review with this much TODOs and other things that are still in progress. I did a partial review with some things that were a bit obvious, other parts I won't assume their status or what happened since I don't know if you'll still address based on what we discussed.

Ktmi commented:
Looking deeper into the code here, the old system wasn't ever making the vlans available when removing the old_path.

Ktmi commented:
For redeploys, its looking like performing the redeploy process in bulk would be too difficult, however, I can get the undeploying of the EVC done, preparing it for a proper redeploy.

Ktmi commented:
Just finished writing some e2e tests for the consistency behaviour. You can find them at kytos-ng/kytos-end-to-end-tests#339. The current version of mef_eline fails all three of those tests. This PR only fails the last one, so I'll be hunting down the cause of that bug.

Ktmi commented:
Here are the E2E results with the new tests:

```
kytos-1  | Starting enhanced syslogd: rsyslogd.
kytos-1  | /etc/openvswitch/conf.db does not exist ... (warning).
kytos-1  | Creating empty database /etc/openvswitch/conf.db.
kytos-1  | Starting ovsdb-server.
kytos-1  | rsyslogd: error during config processing: omfile: chown for file '/var/log/syslog' failed: Operation not permitted [v8.2302.0 try https://www.rsyslog.com/e/2207 ]
kytos-1  | Configuring Open vSwitch system IDs.
kytos-1  | Starting ovs-vswitchd.
kytos-1  | Enabling remote OVSDB managers.
kytos-1  | + '[' -z '' ']'
kytos-1  | + '[' -z '' ']'
kytos-1  | + echo 'There is no NAPPS_PATH specified. Default will be used.'
kytos-1  | + NAPPS_PATH=
kytos-1  | There is no NAPPS_PATH specified. Default will be used.
kytos-1  | + sed -i 's/STATS_INTERVAL = 60/STATS_INTERVAL = 7/g' /var/lib/kytos/napps/kytos/of_core/settings.py
kytos-1  | + sed -i 's/CONSISTENCY_MIN_VERDICT_INTERVAL =.*/CONSISTENCY_MIN_VERDICT_INTERVAL = 60/g' /var/lib/kytos/napps/kytos/flow_manager/settings.py
kytos-1  | + sed -i 's/LINK_UP_TIMER = 10/LINK_UP_TIMER = 1/g' /var/lib/kytos/napps/kytos/topology/settings.py
kytos-1  | + sed -i 's/DEPLOY_EVCS_INTERVAL = 60/DEPLOY_EVCS_INTERVAL = 5/g' /var/lib/kytos/napps/kytos/mef_eline/settings.py
kytos-1  | + sed -i 's/LLDP_LOOP_ACTIONS = \["log"\]/LLDP_LOOP_ACTIONS = \["disable","log"\]/' /var/lib/kytos/napps/kytos/of_lldp/settings.py
kytos-1  | + sed -i 's/LLDP_IGNORED_LOOPS = {}/LLDP_IGNORED_LOOPS = {"00:00:00:00:00:00:00:01": \[\[4, 5\]\]}/' /var/lib/kytos/napps/kytos/of_lldp/settings.py
kytos-1  | + sed -i 's/CONSISTENCY_COOKIE_IGNORED_RANGE =.*/CONSISTENCY_COOKIE_IGNORED_RANGE = [(0xdd00000000000000, 0xdd00000000000009)]/g' /var/lib/kytos/napps/kytos/flow_manager/settings.py
kytos-1  | + sed -i 's/LIVENESS_DEAD_MULTIPLIER =.*/LIVENESS_DEAD_MULTIPLIER = 3/g' /var/lib/kytos/napps/kytos/of_lldp/settings.py
kytos-1  | + kytosd --help
kytos-1  | + sed -i s/WARNING/INFO/g /etc/kytos/logging.ini
kytos-1  | + test -z ''
kytos-1  | + TESTS=tests/
kytos-1  | + test -z ''
kytos-1  | + RERUNS=2
kytos-1  | + python3 scripts/wait_for_mongo.py
kytos-1  | Trying to run hello command on MongoDB...
kytos-1  | Trying to run 'hello' command on MongoDB...
kytos-1  | Trying to run 'hello' command on MongoDB...
kytos-1  | Ran 'hello' command on MongoDB successfully. It's ready!
kytos-1  | + python3 -m pytest tests/test_e2e_17_mef_eline.py
kytos-1  | ============================= test session starts ==============================
kytos-1  | platform linux -- Python 3.11.2, pytest-8.1.1, pluggy-1.5.0
kytos-1  | rootdir: /tests
kytos-1  | plugins: rerunfailures-13.0, timeout-2.2.0, anyio-4.3.0
kytos-1  | collected 3 items
kytos-1  | 
kytos-1  | tests/test_e2e_17_mef_eline.py ..F                                       [100%]
kytos-1  | 
kytos-1  | =================================== FAILURES ===================================
kytos-1  | ______________________ TestE2EMefEline.test_003_link_down ______________________
kytos-1  | 
kytos-1  | self = <tests.test_e2e_17_mef_eline.TestE2EMefEline object at 0x7f70779f8290>
kytos-1  | 
kytos-1  |     def test_003_link_down(self):
kytos-1  |         """Test link down behaviour on failover_path."""
kytos-1  |     
kytos-1  |         self.net.net.configLinkStatus("s1", "s6", "down")
kytos-1  |         self.net.net.configLinkStatus("s3", "s6", "down")
kytos-1  |         self.net.net.configLinkStatus("s5", "s6", "down")
kytos-1  |     
kytos-1  |         payload = {
kytos-1  |             "name": "Link Down Test",
kytos-1  |             "uni_a": {"interface_id": "00:00:00:00:00:00:00:01:1", "tag": {"tag_type": "vlan", "value": 100}},
kytos-1  |             "uni_z": {"interface_id": "00:00:00:00:00:00:00:05:1", "tag": {"tag_type": "vlan", "value": 100}},
kytos-1  |             "enabled": True,
kytos-1  |             "primary_constraints": {
kytos-1  |                 "mandatory_metrics": {
kytos-1  |                     "not_ownership": ["forbidden_link"],
kytos-1  |                 },
kytos-1  |             },
kytos-1  |             "dynamic_backup_path": True,
kytos-1  |         }
kytos-1  |         api_url = KYTOS_API + "/mef_eline/v2/evc/"
kytos-1  |         response = requests.post(api_url, json=payload)
kytos-1  |     
kytos-1  |         assert response.status_code == 201, response.text
kytos-1  |     
kytos-1  |         data = response.json()
kytos-1  |         evc_id =  data["circuit_id"]
kytos-1  |     
kytos-1  |         time.sleep(10)
kytos-1  |     
kytos-1  |         api_url = KYTOS_API + "/mef_eline/v2/evc/"
kytos-1  |         response = requests.get(api_url + evc_id)
kytos-1  |         data = response.json()
kytos-1  |     
kytos-1  |         assert data["current_path"]
kytos-1  |         assert data["failover_path"]
kytos-1  |     
kytos-1  |         # Collect service vlans
kytos-1  |     
kytos-1  |         vlan_allocations = defaultdict[str, list[int]](list)
kytos-1  |     
kytos-1  |         for link in data["failover_path"]:
kytos-1  |             s_vlan = link["metadata"]["s_vlan"]
kytos-1  |             for endpoint in (link["endpoint_a"], link["endpoint_b"]):
kytos-1  |                 vlan_allocations[endpoint["id"]].append(s_vlan)
kytos-1  |     
kytos-1  |         # Close a link that the failover path depends on
kytos-1  |     
kytos-1  |         link = data["failover_path"][1]
kytos-1  |         if link["id"] == LinkID("00:00:00:00:00:00:00:02:3", "00:00:00:00:00:00:00:03:2"):
kytos-1  |             self.net.net.configLinkStatus("s2", "s3", "down")
kytos-1  |         else:
kytos-1  |             self.net.net.configLinkStatus("s2", "s6", "down")
kytos-1  |     
kytos-1  |         time.sleep(10)
kytos-1  |     
kytos-1  |         # EVC should be enabled but not active
kytos-1  |     
kytos-1  |         api_url = KYTOS_API + "/mef_eline/v2/evc/"
kytos-1  |         response = requests.get(api_url + evc_id)
kytos-1  |         data = response.json()
kytos-1  |     
kytos-1  |         assert data["enabled"]
kytos-1  |         assert data["active"]
kytos-1  |     
kytos-1  |         assert data["current_path"]
kytos-1  |         assert not data["failover_path"]
kytos-1  |     
kytos-1  |         # Check that all the s_vlans have been freed
kytos-1  |     
kytos-1  |         api_url = f"{KYTOS_API}/topology/v3/interfaces/tag_ranges"
kytos-1  |     
kytos-1  |         response = requests.get(api_url)
kytos-1  |     
kytos-1  |         assert response.ok, response.text
kytos-1  |     
kytos-1  |         data = response.json()
kytos-1  |     
kytos-1  |         for interface, reserved_tags in vlan_allocations.items():
kytos-1  |             available_tags = data[interface]["available_tags"]
kytos-1  |             for reserved_tag in reserved_tags:
kytos-1  | >               assert any(
kytos-1  |                     reserved_tag["value"] >= range_start and reserved_tag["value"] <= range_end
kytos-1  |                     for (range_start, range_end) in available_tags[reserved_tag["tag_type"]]
kytos-1  |                 ), f"Vlan tag {reserved_tag} on interface {interface}, not released. Available tags: {available_tags}"
kytos-1  | E               AssertionError: Vlan tag {'tag_type': 'vlan', 'value': 2} on interface 00:00:00:00:00:00:00:01:2, not released. Available tags: {'vlan': [[3, 3798], [3800, 4095]]}
kytos-1  | E               assert False
kytos-1  | E                +  where False = any(<generator object TestE2EMefEline.test_003_link_down.<locals>.<genexpr> at 0x7f70779297e0>)
kytos-1  | 
kytos-1  | tests/test_e2e_17_mef_eline.py:394: AssertionError
kytos-1  | =============================== warnings summary ===============================
kytos-1  | test_e2e_17_mef_eline.py: 37 warnings
kytos-1  |   /usr/lib/python3/dist-packages/mininet/node.py:1121: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
kytos-1  |     return ( StrictVersion( cls.OVSVersion ) <
kytos-1  | 
kytos-1  | test_e2e_17_mef_eline.py: 37 warnings
kytos-1  |   /usr/lib/python3/dist-packages/mininet/node.py:1122: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
kytos-1  |     StrictVersion( '1.10' ) )
kytos-1  | 
kytos-1  | -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
kytos-1  | ------------------------------- start/stop times -------------------------------
kytos-1  | test_e2e_17_mef_eline.py::TestE2EMefEline::test_003_link_down: 2024-12-13,17:08:45.783771 - 2024-12-13,17:09:05.906983
kytos-1  | =========================== short test summary info ============================
kytos-1  | FAILED tests/test_e2e_17_mef_eline.py::TestE2EMefEline::test_003_link_down - ...
kytos-1  | ============= 1 failed, 2 passed, 74 warnings in 155.08s (0:02:35) =============
```

Ktmi commented:
@viniarck This should be ready for a final review. Two things of note: one of the tests still needs a rewrite, as we added additionally functionality to link down handler that existing tests don't implement mocks for. Additionally, using the http endpoint for sending the flow mods didn't work out. E2E tests started to fail with that, so I reverted that change.

Ktmi commented:
Performed a rebase to make it easier to cherry pick commits.

Alopalao commented:
LGTM, I did some testing to see if the EVC locks are hold and seems to be working as intended. Great reworking of old_path management.

italovalcy commented:
Hi David, I tested this PR using some of the scenarios we saw recently in production and all worked as expected. Very nice done! LGTM!

viniarck commented on main.py:
```diff
@@ -823,85 +825,289 @@ def on_link_down(self, event):
 
-        emit_event(
-            self.controller,
-            "cleanup_evcs_old_path",
```
7) remove code that no longer is used [`on_cleanup_evcs_old_path` handler](https://github.com/kytos-ng/mef_eline/blob/fix/cleanup_old_path/main.py#L1147) that's subscribed to `kytos/mef_eline.cleanup_evcs_old_path` (and remove it from the README.rst too)

Ktmi replied:
Removed

viniarck reviewed (changes requested):
Great fix and overall improvement compared to what we used to have. Also, neat use of multi lock with nested `ExitStack`s.

It's almost there, I'm requesting changes since there's a major KytosEvent breaking change on `telemetry_int` item 6) below, and the thread safety issue on item 3) is a bit concerning too, although it'd be very unlikely to happen, but it can happen, there's a risk involved that we need to make a decision. Also, unit tests needs to be covered.

Ktmi commented on main.py:
```diff
@@ -823,85 +825,289 @@ def on_link_down(self, event):
-            content={"evcs": evcs_with_failover + check_failover}
-        )
+            if clear_old_path_flows:
+                self.execute_clear_old_path(
```
Trying the E2E tests with the bulk flows resulted in several regressions in E2E tests, so I decided to revert that change.

viniarck replied:
Interesting. Do you recall if you tried to narrow down what was the underlying cause? If it was something on flow_manager that deleted more or less flows than expected? Because that's supposed to be equivalent except in a single request while doing all the ops in the request response cycle. E2e coverage for that endpoint isn't too through though but we [have some coverage](https://github.com/kytos-ng/kytos-end-to-end-tests/pull/331/files).

Another pro about it too, is when we allow direct NApp to NApp func calls we can get rid of the request overhead too, assuming no other bugs that you mentioned.

Ktmi replied:
Also here's the patch file that's being applied to mef_eline.
```patch
diff --git a/main.py b/main.py
index 0b7a1a3..cd6edf7 100755
--- a/main.py
+++ b/main.py
@@ -34,7 +34,7 @@ from napps.kytos.mef_eline.utils import (aemit_event, check_disabled_component,
                                          emit_event, get_vlan_tags_and_masks,
                                          map_evc_event_content,
                                          merge_flow_dicts, prepare_delete_flow,
-                                         send_flow_mods_event)
+                                         send_flow_mods_http)
 
 
 # pylint: disable=too-many-public-methods
@@ -848,7 +848,7 @@ class Main(KytosNApp):
             self.controller, "failover_link_down",
             content=deepcopy(event_contents)
         )
-        send_flow_mods_event(
+        send_flow_mods_http(
             self.controller,
             install_flows,
             "install"
@@ -877,7 +877,7 @@ class Main(KytosNApp):
         delete_flows
     ):
         """Process changes needed to commit clearing the old path"""
-        send_flow_mods_event(
+        send_flow_mods_http(
             self.controller,
             delete_flows,
             'delete'
@@ -911,7 +911,7 @@ class Main(KytosNApp):
 
     def execute_undeploy(self, evcs: list[EVC], remove_flows):
         """Process changes needed to commit an undeploy"""
-        send_flow_mods_event(
+        send_flow_mods_http(
             self.controller,
             remove_flows,
             'delete'
diff --git a/utils.py b/utils.py
index 8ae4e97..bfe23ee 100644
--- a/utils.py
+++ b/utils.py
@@ -4,8 +4,10 @@ from typing import Union
 from kytos.core.common import EntityStatus
 from kytos.core.events import KytosEvent
 from kytos.core.interface import UNI, Interface, TAGRange
-from napps.kytos.mef_eline.exceptions import DisabledSwitch
+from napps.kytos.mef_eline import settings
+from napps.kytos.mef_eline.exceptions import DisabledSwitch, FlowModException
 
+import httpx
 
 def map_evc_event_content(evc, **kwargs) -> dict:
     """Returns a set of values from evc to be used for content"""
@@ -178,6 +180,35 @@ def send_flow_mods_event(
             },
         )
 
+def send_flow_mods_http(
+    controller,
+    flow_dict: dict[str, list],
+    action: str, force=True
+):
+    """
+    Send a flow_mod list to a specific switch.
+
+    Args:
+        dpid(str): The target of flows (i.e. Switch.id).
+        flow_mods(dict): Python dictionary with flow_mods.
+        command(str): By default is 'flows'. To remove a flow is 'remove'.
+        force(bool): True to send via consistency check in case of errors.
+        by_switch(bool): True to send to 'flows_by_switch' request instead.
+    """
+    endpoint = f"{settings.MANAGER_URL}/flows_by_switch/?force={force}"
+
+    try:
+        if action == "install":
+            res = httpx.post(endpoint, json=flow_dict, timeout=30)
+        elif action == "delete":
+            res = httpx.request(
+                "DELETE", endpoint, json=flow_dict, timeout=30
+            )
+    except httpx.RequestError as err:
+        raise FlowModException(str(err)) from err
+    if res.is_server_error or res.status_code >= 400:
+        raise FlowModException(res.text)
+
 
 def prepare_delete_flow(evc_flows: dict[str, list[dict]]):
     """Create flow mods suited for flow deletion."""

```

viniarck replied:
@Ktmi, let's get to the bottom of it. Also, by reading the patch and the temporary refactoring, it looks like it's passing an invalid payload to the bulk endpoint `{dpid: [flows]}` instead of `{dpid: {flows: [flows]}}`. Can you double check this, and confirmed if you ran it locally? 

> However, I don't know how to get the logs from kytos itself while running e2e tests. If you can provide some instructions on how to do so, I could provide additional context.

The quickest way is to read from `kytos.log`, you have to add the `file` handler on the root [logging.ini config](https://github.com/kytos-ng/kytos/blob/master/kytos/templates/logging.ini.template#L37). Ultimately, you need to correlate the timestamp and see start and end times, that's how we debug e2e tests on both CI and locally typically. 

If you're running `kytosd` in a Docker container (which seems the case), then the easiest route after making sure the file handler is set, is to `tail -f /var/log/kytos.log` in another terminal window/panel (or add another command in the shell script to copy the file before the container exists). 

If you're running e2e without `kytosd` in a container it's easier since `kytos.log` won't be removed automatically, so it's easier to read after when the test fails.

Ktmi commented on main.py:
```diff
@@ -823,85 +825,289 @@ def on_link_down(self, event):
+                            flows=deepcopy(new_flows)
+                    )
+                    clear_old_path.append(evc)
+                elif evc.id in flow_modifications:
```
It is unreachable, but left in as part of mentally mapping how the procedure should go. And now that I look at it, it still has a bug, as the undeploy should only be conditional on the failure to generate flows, not on if the flow_modifications needed to be removed.

viniarck replied:
Right.

Ktmi replied:
The logic bug here should be fixed, and I removed the check on the flow modifications in this particular case.

Ktmi commented on main.py:
```diff
@@ -823,85 +825,289 @@ def on_link_down(self, event):
+                return
+            result = evc.deploy_to_path()
+            # if result:
+            #     evc.setup_failover_path()
```
The intention was indeed to try to trigger setup failover, but that doesn't work due to `try_setup_failover_path()` holding the evc lock.

viniarck replied:
Cool

Ktmi replied:
Removed the comment

Ktmi commented on main.py:
```diff
@@ -823,85 +825,289 @@ def on_link_down(self, event):
-                    # tradditional way
-                    if (
+        with ExitStack() as exit_stack:
+            exit_stack.enter_context(self.multi_evc_lock)
```
I did it to ensure we have a pattern in place here. If we decide to add other instances where we modify multiple EVCs in a large batch like this, then we won't have to go back and modify this code, only include this pattern in the new code. If you do try to acquire multiple EVC locks without a general lock for the ability to hold multiple locks, then you risk a dead lock.

viniarck replied:
I see, indeed it's good to already keep this this pattern in place, even though now it wouldn't dead lock. Let's follow this pattern then from now on mef_eline for any other bulk EVC ops; closing this thread comment here.

Ktmi commented on main.py:
```diff
@@ -823,85 +825,289 @@ def on_link_down(self, event):
+                    event_contents[evc.id]["clear_old_path"] =\
+                        map_evc_event_content(
+                            evc,
+                            removed_flows=deepcopy(del_flows)
```
Added back in, will try to test in the int lab later.

Ktmi replied:
Ran this on the lab today. telemetry_int appears to now be properly reacting to the the failover changes.

Ktmi commented on tests/unit/test_main.py:
```diff
@@ -1899,6 +1899,9 @@ def test_handle_link_up(self):
         assert evc_mock.handle_link_up.call_count == 2
         evc_mock.handle_link_up.assert_called_with("abc")
 
+    @pytest.mark.xfail(
```
Unit tests updated.

viniarck commented on utils.py:
```diff
@@ -179,6 +182,40 @@ def send_flow_mods_event(
         )
 
 
+def send_flow_mods_http(
```
Highly recommend that you decorate it with tenacity's `@retry` similar pattern that we have [on other NApps](https://github.com/kytos-ng/telemetry_int/blob/master/kytos_api_helper.py#L22), if the http server is temporarily overloaded this can help to succeed in certain transient cases

Ktmi replied:
Tenacity has been added to the endpoint.

viniarck commented on main.py:
```diff
@@ -833,85 +835,284 @@ def on_link_down(self, event):
+                evcs_to_update.append(evc)
+
+            if swap_to_failover_flows:
+                self.execute_swap_to_failover(
```
All `execute_*` calls which will ultimately synchronously call `send_flow_mods_http(*args)`, when `send_flow_mods_http` fails, this will fail in a single thread context, will be sent to the dead queue letter, and will leave in a situation which a highly likely broken data plane as expected. That said, if it happens, it'll be stressful in prod to deal with, and http server is usually more likely to get overloaded and fail than the DB cluster on avg, what I recommend you to do and consider is to try catch here and: 

1) log the EVCs as an error, if it blows then at least there will be clear log for you to also to communicate and orient the Ops team to understand what's going on, in this log it's probably also worth recommending to reploy in a next opportunity until we have other mechanisms that can also try to recover from it (or if no other related events have taken place then they also can reinject the event in the queue).
2) and then reraise and let it crash just so it ends up in the dead queue letter. 

Finally, make sure to exercise locally this http failure here, and see if you're happy supporting it if it ever goes south in an unlikely rare event when things are overloaded, if you are, then ship it.

Ktmi replied:
I redid this. I think I have it in a state where it will be gracefully handled.

viniarck reviewed (approved):
Great to have the improvements and this issue fixed @Ktmi. 

Left a suggestion to include http retries, and also recommended another log if you wish, up to you. Other than that, go ahead and ship it and let the Ops team know when the related backport is also ready to be used in prod.

viniarck commented on main.py:
```diff
@@ -833,85 +836,272 @@ def on_link_down(self, event):
+
+            # Undeploy the evc, schedule a redeploy
+
+            success, failure = self.execute_undeploy(undeploy)
```
Should we log the `failure` list as error after this statement if it's not empty? Since after this statement we're not doing anything with the `failure` list. The `prepare_*` errors will be extremely rare, but if they do happen, would be cool to also have a log error for us to debug and understand in the future if it ever happens.

italovalcy replied:
+1 for having more log statements in case of failure (of course, we have to measure the compromise with performance impact, but one log message at least is beneficial).

Ktmi replied:
Logging has been added in. Also E2E tests are now passing.

viniarck reviewed (approved):
LGTM @Ktmi, just did another pass. Great to have the improvements and this fixed. 

@italovalcy is planning to also deploy the backport version of this PR, if e2e tests are passing let's plan to land this and the backport one, also ping Italo to see if might want to re-review or quick explore something else since he was also involved a lot in this issue/PR otherwise let's ship it.
